File: standalone/examples/primal_demo.py
#!/usr/bin/env python3
"""Minimal demo + global-merge snippet for the standalone primal port.
    python3 primal_demo.py
"""
import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "python"))
import subgrid_mt as smt  # noqa: E402


def main():
    # One tet with a few intersections on its edges.
    pos = [smt.Vec3(0, 0, 0), smt.Vec3(1, 0, 0), smt.Vec3(0, 1, 0), smt.Vec3(0, 0, 1)]
    gidx = [0, 1, 2, 3]
    ts = [[] for _ in range(6)]  # edges (0,1),(0,2),(0,3),(1,2),(1,3),(2,3)
    # A single loop cutting off vertex 0 (crosses the 3 edges incident to it):
    # every face is even-sum, so this is a normal curve -> one triangle.
    ts[0] = [0.3]
    ts[1] = [0.3]
    ts[2] = [0.3]

    r = smt.subgrid_primal(pos, gidx, ts)
    print(f"vertices: {len(r.vertices)}, faces: {len(r.faces)}, non_even: {r.non_even}")

    # ---- Global cross-tet merge sketch (mirrors add_local_soup_comb) ----
    # Accumulate a PrimalResult for every active tet, then merge by signature.
    global_v = []
    global_f = []
    key_to_gid = {}

    def add_result(res):
        local_to_global = [0] * len(res.vertices)
        for vi in range(len(res.vertices)):
            key = smt.signature_key(res.signatures[vi])
            if key is None:  # STANDALONE -> always fresh
                local_to_global[vi] = len(global_v)
                global_v.append(res.vertices[vi])
            elif key in key_to_gid:
                local_to_global[vi] = key_to_gid[key]
            else:
                local_to_global[vi] = len(global_v)
                key_to_gid[key] = len(global_v)
                global_v.append(res.vertices[vi])
        for f in res.faces:
            global_f.append([local_to_global[idx] for idx in f])

    add_result(r)
    print(f"merged global vertices: {len(global_v)}, faces: {len(global_f)}")


if __name__ == "__main__":
    main()

File: standalone/examples/primal_demo.cpp
// Minimal demo + global-merge snippet for the standalone primal port.
//   c++ -std=c++17 -I../cpp primal_demo.cpp -o primal_demo && ./primal_demo
#include "subgrid_mt.hpp"

#include <iostream>
#include <unordered_map>

using namespace subgrid_mt;

int main() {
    // One tet with a few intersections on its edges.
    array<Vec3, 4> pos = {Vec3{0, 0, 0}, Vec3{1, 0, 0}, Vec3{0, 1, 0}, Vec3{0, 0, 1}};
    array<size_t, 4> gidx = {0, 1, 2, 3};
    array<vector<double>, 6> ts;  // edges (0,1),(0,2),(0,3),(1,2),(1,3),(2,3)
    // A single loop cutting off vertex 0 (crosses the 3 edges incident to it):
    // every face is even-sum, so this is a normal curve -> one triangle.
    ts[0] = {0.3};
    ts[1] = {0.3};
    ts[2] = {0.3};

    PrimalResult r = subgrid_primal(pos, gidx, ts);
    std::cout << "vertices: " << r.vertices.size() << ", faces: " << r.faces.size()
              << ", non_even: " << (r.non_even ? "true" : "false") << "\n";

    // ---- Global cross-tet merge sketch (mirrors add_local_soup_comb) ----
    // Accumulate PrimalResult r for every active tet, then:
    std::vector<Vec3> global_v;
    std::vector<std::vector<size_t>> global_f;
    std::unordered_map<std::string, size_t> key_to_gid;
    auto add_result = [&](const PrimalResult& res) {
        std::vector<size_t> local_to_global(res.vertices.size());
        for (size_t vi = 0; vi < res.vertices.size(); vi++) {
            std::string key = signature_key(res.signatures[vi]);
            if (key.empty()) {  // STANDALONE -> always fresh
                local_to_global[vi] = global_v.size();
                global_v.push_back(res.vertices[vi]);
            } else {
                auto it = key_to_gid.find(key);
                if (it != key_to_gid.end()) {
                    local_to_global[vi] = it->second;
                } else {
                    local_to_global[vi] = global_v.size();
                    key_to_gid[key] = global_v.size();
                    global_v.push_back(res.vertices[vi]);
                }
            }
        }
        for (const auto& f : res.faces) {
            std::vector<size_t> gf;
            for (size_t idx : f) gf.push_back(local_to_global[idx]);
            global_f.push_back(gf);
        }
    };
    add_result(r);
    std::cout << "merged global vertices: " << global_v.size() << ", faces: " << global_f.size() << "\n";
    return 0;
}
